### PR TITLE
NChooseK constraints in Samplers

### PR DESCRIPTION
**Pull request by jduerholt**
This PR adds the following things:
- NChooseK sampling implemented for `PolytopeSampler and `RejectionSampler`. Interesting for @dlinzner-bcs 
- First steps of continuous relaxation of `NChooseK` constraints for BayesOpt

### CODE REVIEW

File: bofire/utils/torch_tools.py [hunk 1]
```diff
@@ -79,6 +80,44 @@ def get_linear_constraints(
...
+            as nonlinear equality constraints in botorch.
+    """
+
+    def narrow_gaussian(x, ell=1e-3):
+        return torch.exp(-0.5 * (x / ell) ** 2)
+
+    constraints = []
+    # ignore none also valid for the start
+    for c in domain.cnstrs.get(NChooseKConstraint):
+        assert isinstance(c, NChooseKConstraint)
```
Comment (jkleinekorte) [inline bofire/utils/torch_tools.py:103]: warum brauchst du das nochmal hier? Sollte doch eig bereits in constraints abgefangen sein

File: notebooks/xt.ipynb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (jkleinekorte) [inline notebooks/xt.ipynb]: ich vermute, das hier sollte eher nicht im Git landen? ;)

**Review by jkleinekorte — APPROVED** [review]

File: bofire/utils/torch_tools.py (hunk 1 above)
Comment (jduerholt) [reply bofire/utils/torch_tools.py:103]: Nur für pyright, damit der nicht meckert.

File: notebooks/xt.ipynb (hunk 2 above)
Comment (jduerholt) [reply notebooks/xt.ipynb]: habe ich schon rausgeworfen.

File: bofire/samplers.py [hunk 3]
```diff
@@ -63,20 +64,58 @@ class Sampler(PydanticBaseModel):
...
         Returns:
             pd.DataFrame: Dataframe with samples.
         """
+        # handle here NChooseK
+        if len(self.domain.cnstrs.get(NChooseKConstraint)) > 0:
+            _, unused = self.domain.get_nchoosek_combinations()
+            samples = []
+            for u in unused:
+                # create new domain without the nchoosekconstraints
+                domain = deepcopy(self.domain)
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/samplers.py:85]: I think this is an overkill as the thing that we want to achieve can be done in just two lines.

File: bofire/domain/domain.py [hunk 4]
```diff
@@ -315,11 +318,16 @@ def get_nchoosek_combinations(self):
...
+                    used_features_list.extend(itertools.combinations(con.features, n))
 
-            if con.none_also_valid:
-                used_features_list.append(tuple([]))
+                if con.none_also_valid:
+                    used_features_list.append(tuple([]))
+            else:
+                used_features_list.extend(
+                    itertools.combinations(con.features, con.max_count)
+                )
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/domain/domain.py:341]: I was not sure, if it makes really sense in the non exhaustive setting. Will think about it when finalizing the feature and adding it to base.py

**Pull request merged**